### DISCUSSION TITLE
Adding A Typescript from C#

### DISCUSSION THREAD

**Discussion by JesseAtRelish** [naverz/zepeto-studio-global#27] [Q&A]
Hello,

Typically in C# we can add scripts to game objects with gameObject.AddComponent(typeof(Node));
How can we add a ZepetoScript to a gameobject?

Cheers,

**Comment by JesseAtRelish** [comment]
This seems to be the answer I was looking for...

gameObject.AddComponent(typeof(ZEPETO.Script.ZepetoScriptBehaviourComponent));

**Comment by JesseAtRelish** [comment]
Adding the ZepetoScriptBehaviourComponents seems like the component I need. But still can't figure out how to add a reference to the typescript..


```
ZEPETO.Script.ZepetoScriptBehaviourComponent comp = gameObject.AddComponent<ZEPETO.Script.ZepetoScriptBehaviourComponent>();

foreach (string types in AssetDatabase.FindAssets("t:ZEPETO.Script.TypescriptAsset")) {
    string path = AssetDatabase.GUIDToAssetPath(types);
    ZEPETO.Script.TypescriptAsset obj = (ZEPETO.Script.TypescriptAsset)AssetDatabase.LoadAssetAtPath(path, typeof(ZEPETO.Script.TypescriptAsset));
     if (path.Contains("ZNode")) {
       // comp.script = ZEPETO.Script.ZepetoScriptInstance  //<------- How to get instance from the TypescriptAsset!!
     }
}
```

**Comment by jhslee** [comment]
Hello @JesseAtRelish 👋 

To add a component to an object, simply use `gameObject.AddComponent<Type>()` and define the type you want to add in the brackets (No need for typeof). Please reference the following lines as an example:

If we have a ball Class:
```
export default class Ball extends ZepetoScriptBehaviour {
    public bounceHeight: number;
}
```

And a seperate script that adds a typescript to objects:
```
import Ball from './Ball'; //please be sure that you add the proper import based on your file structure

export default class AddComponentTest extends ZepetoScriptBehaviour {

     public Start()
     {
          const ball = gameObject.AddComponent<Ball>();
          ball.bounceHeight = 10;
     }
     
}
```

**Comment by JesseAtRelish** [comment]
This was my first approach. However the extended Ball class (in this case)
doesn't appear as a valid type in the Unity project. We're trying to add a typescript class from c#.

On Wed, Aug 3, 2022 at 12:19 PM Zepeto-TechEvangelist <
***@***.***> wrote:

> Hello @JesseAtRelish <https://github.com/JesseAtRelish> 👋
>
> To add a component to an object, simply use
> gameObject.AddComponent<Type>() and define the type you want to add in
> the brackets (No need for typeof). Please reference the following lines as
> an example:
>
> If we have a ball Class:
>
> export default class Ball extends ZepetoScriptBehaviour {
>
>     public bounceHeight: number;
>
> }
>
>
> And a seperate script that adds a typescript to objects:
>
> const ball = gameObject.AddComponent<Ball>();
>
> ball.bounceHeight = 10;
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/naverz/zepeto-studio-global/discussions/27#discussioncomment-3318049>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANTXISX2QCWXXNFENEUP2QLVXKLX3ANCNFSM55LJC47Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 

Jesse LaChapelle (he/him)

Storylines Transition Coordinator

***@***.***

**Reply by jhslee** [reply]
You will need to ensure that the type is imported into that typescript file so that the script can see which type you're referencing. 

Most IDEs (Like visual studio code for example) will give you an error in the line when you try to reference an external type. 
<img width="629" alt="Screen Shot 2022-08-03 at 9 27 53 AM" src="https://user-images.githubusercontent.com/105688184/182660807-23f2d0b0-cc16-434f-ae06-38f4cecbd338.png">

However usually they'll give you the option to resolve that error by adding in the reference automatically:
<img width="841" alt="Screen Shot 2022-08-03 at 9 28 03 AM" src="https://user-images.githubusercontent.com/105688184/182660981-dcbfc047-b1d5-473b-bb08-0760f2bef029.png">

If the import reference is correctly defined, the errors should dissapear.

**Comment by JesseAtRelish** [comment]
Firstly, thank you for your assistance. Much appreciated.

I think you're suggesting to add my custom ZepetoBehaviour (ie. Ball), to
the import of another typescript. We're trying to use some existing editor
tools written in C#, and we'd like to extend it to add the required
typescript files. The custom typescript doesn't appear as a valid type in
our C# code.

On Wed, Aug 3, 2022 at 12:30 PM Zepeto-TechEvangelist <
***@***.***> wrote:

> You will need to ensure that the type is imported into that typescript
> file so that the script can see which type you're referencing.
>
> Most IDEs (Like visual studio code for example) will give you an error in
> the line when you try to reference an external type.
> [image: Screen Shot 2022-08-03 at 9 27 53 AM]
> <https://user-images.githubusercontent.com/105688184/182660807-23f2d0b0-cc16-434f-ae06-38f4cecbd338.png>
>
> However usually they'll give you the option to resolve that error by
> adding in the reference automatically:
> [image: Screen Shot 2022-08-03 at 9 28 03 AM]
> <https://user-images.githubusercontent.com/105688184/182660981-dcbfc047-b1d5-473b-bb08-0760f2bef029.png>
>
> If the import reference is correctly defined, the errors should dissapear.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/naverz/zepeto-studio-global/discussions/27#discussioncomment-3318137>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANTXISQMRRXB7XU4IJT47XTVXKNEBANCNFSM55LJC47Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 

Jesse LaChapelle (he/him)

Storylines Transition Coordinator

***@***.***

**Reply by jhslee** [reply]
ah~~ I see. Ok got it, my apologies, I didn't realize that context 😅. I actually don't believe our typescript compiler processes the typescript classes in a way that makes them visible to the C# compiler. I'll double check on this internally with our devs, and see if this is something that can be done. 😁 

I noticed you said the C# code will be for your in-editor tools so this may be a non-issue, but I just wanted to clarify that C# code will not run on the device, even if it may seem to work from the editor. Any C# code will be ignored when testing from within the zepeto application.

**Comment by JesseAtRelish** [comment]
The tools generate a gameobject hierarchy that defines our game. These game
objects have C# editor scripts... my plan was to add parallel typescript
scripts to the hierarchy and use the editor to configure them. At runtime
the C# versions are not used, just the configured typescripts.

I see there are some C# scripts maybe bundled in a dll (ZepetoCharacter,
ZepetoCamera for instance), that seem to be able to make use of typescript
classes, so I'm hopeful there's some way to extend the editor to interact
with the typescript layer.

This is why I was adding 'ZepetoScript' to the gameobjects. Then looking to
add a ZepetoScriptInstance to the 'script' field (I think).

On Wed, Aug 3, 2022 at 12:56 PM Zepeto-TechEvangelist <
***@***.***> wrote:

> ah~~ I see. Ok got it, my apologies, I didn't realize that context 😅. I
> actually don't believe our typescript compiler processes the typescript
> classes in a way that makes them visible to the C# compiler. I'll double
> check on this internally with our devs, and see if this is something that
> can be done. 😁
>
> I noticed you said the C# code will be for your in-editor tools so this
> may be a non-issue, but I just wanted to clarify that C# code will not run
> on the device, even if it may seem to work from the editor. Any C# code
> will be ignored when testing from within the zepeto application.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/naverz/zepeto-studio-global/discussions/27#discussioncomment-3318335>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANTXIST6ZBTCTF2FIVEOAFDVXKQDFANCNFSM55LJC47Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 

Jesse LaChapelle (he/him)

Storylines Transition Coordinator

***@***.***

**Reply by jhslee** [reply]
oh wow that's such a cool idea, thank you so much for that clarification! One caveat I'd throw in there is that typescript vs C# may not be a 1:1 translation, something we're noticing lately is that there are several Unity C# library features that are buggy or yet to be implemented to their typescript counterparts. I'm unfortunately unable to give a comprehensive list of all of these missing features, but i'd be mindful of having a backup for your development workflow in case this doesn't work properly, or puts your delivery timeline at risk while waiting for internal devs to implement this feature.

I'll look into this a bit for you, and get back to you as soon as I can!
